### PR TITLE
[cxx-interop] Fix wrong field offsets with value types using inheritance

### PR DESCRIPTION
**Pull request by Xazax-hun**
Instead of adding opaque fields for the base subobjects this patch introduces a recursive walk to add all the base fields to the generated Swift struct.

rdar://126754931

### CODE REVIEW

**Comment by Xazax-hun** [comment]
@swift-ci please smoke test

File: lib/IRGen/GenStruct.cpp [hunk 1]
```diff
@@ -1419,41 +1430,41 @@ class ClangRecordLowering {
...
+    if (NextOffset < objectTotalStride) {
+      addPaddingField(objectTotalStride);
     }
   }
 
   /// Place the next struct field at its appropriate offset.
   void addStructField(const clang::FieldDecl *clangField,
-                      VarDecl *swiftField) {
-    unsigned fieldOffset = ClangLayout.getFieldOffset(clangField->getFieldIndex());
+                      VarDecl *swiftField, const clang::ASTRecordLayout& layout) {
```
Comment (rjmccall) [inline lib/IRGen/GenStruct.cpp:1464]:
```suggestion
                      VarDecl *swiftField, const clang::ASTRecordLayout &layout) {
```

File: lib/IRGen/GenStruct.cpp [hunk 2]
```diff
@@ -1375,17 +1379,24 @@ class ClangRecordLowering {
...
+        SubobjectAdjustment -= baseOffset;
       }
     }
   }
 
-  void collectStructFields() {
-    auto cfi = ClangDecl->field_begin(), cfe = ClangDecl->field_end();
+  void collectStructFields(const clang::RecordDecl *decl) {
+    auto cfi = decl->field_begin(), cfe = decl->field_end();
+    const auto& layout = ClangContext.getASTRecordLayout(decl);
```
Comment (rjmccall) [inline lib/IRGen/GenStruct.cpp:1393]:
```suggestion
    const auto &layout = ClangContext.getASTRecordLayout(decl);
```

**Review by rjmccall — COMMENTED** [review]
LGTM; just two style nits.

**Comment by Xazax-hun** [comment]
@swift-ci please smoke test

**Comment by Xazax-hun** [comment]
There seems to be an unrelated failure on Linux, rerunning.

@swift-ci please smoke test linux

**Comment by Xazax-hun** [comment]
@swift-ci please smoke test linux

**Pull request merged**